backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/directory/impl/LocalHeapDirectoryHolder.java

```diff
@@ -20,17 +20,36 @@ final class LocalHeapDirectoryHolder implements DirectoryHolder {
 
 	private Directory directory;
 
+	private Directory directoryFromPreviousState;
+	private boolean transferred = false;
+
 	LocalHeapDirectoryHolder(LockFactory lockFactory) {
 		this.lockFactory = lockFactory;
 	}
 
+	@Override
+	public void preStart(DirectoryHolder previousDirectoryHolder) {
+		if ( !( previousDirectoryHolder instanceof LocalHeapDirectoryHolder ) ) {
```

Comment:
That such a thing could even happen seems... concerning? If we support such configuration change, we'll have to test it...

Don't you think it would be simpler to avoid handling state saving in `DirectoryHolder`? We could just reuse the whole `DirectoryHolder` in the shard.